### DISCUSSION TITLE
have a single client parameter http_version instead of http1 and http2

### DISCUSSION THREAD

**Discussion by lewoudar** [encode/httpx#1826] [Ideas] [answered]
Hi folks,
I think it would be a great idea to simplify how to designate the http version we want to work with. Currently clients have two arguments `http1` and `http2`. The problem with this is that when we want to support other versions like `http3`, we will have to add additional parameters.
With a single parameter `http_version` taking values like `h1`, `h2` we will reduce (slightly) the number of arguments passed to the client.

**Comment by lovelydinosaur** [comment, marked as answer]
We do have some users who want to be able to specify "force HTTP2", while others who want "HTTP2 if it's available, otherwise HTTP1". I think we'll probably need to stick with separate flags in order to accommodate those kind of different options.

**Comment by lewoudar** [comment]
oh I see, so `http1=True` and `http2=True` is a desired scenario where httpw will first try http2 connection and fallback to http1 if http2 is not supported, right?

**Reply by lovelydinosaur** [reply]
Yup. In the general case you can't know in advance which HTTP versions the server will support.

(Although the "is this http2 or http1" all happens as part of the SSL negotiation, rather than having to attempt one thing first, and then try the other.)

**Reply by lewoudar** [reply]
Thanks for the answer
